### ISSUE TITLE
Does BYOR support csv content served from S3

### ISSUE THREAD

**Issue by ab73863** [thoughtworks/build-your-own-radar#66]
Hi ,

Could you please let me if I can use S3 storage of a CSV file and use it in BYOR ?

Regards
Arun

**Comment by sgarap** [comment]
Loading the CSV file from S3 is not working. It results in the following error. Can you  please support this feature?

> Oops! It seems like there are some problems with loading your data. Document is missing one or more required headers or they are misspelled. Check that your document contains headers for "name", "ring", "quadrant", "isNew", "description".
> Please check FAQs for possible solutions.

Regards,
Sindhura

**Comment by SARAN-thala** [comment]
Hi @sgarap 

> I tried also with the CSV data from gist https://gist.githubusercontent.com/sgarap/32c02feeec4fc010a332ac7013559e38/raw/d1d0a1575bd47a0502fabe0c94909e99a8039571/test.csv
using the sample data provided in https://github.com/thoughtworks/build-your-own-radar.
The result is the same. Can some one help with loading a CSV data from somewhere, preferably from S3.
It's created the radar for me. please check once again.

**Comment by sgarap** [comment]
@SARAN-thala I tried again with input https://gist.githubusercontent.com/sgarap/32c02feeec4fc010a332ac7013559e38/raw/d1d0a1575bd47a0502fabe0c94909e99a8039571/test.csv

I used Mozilla browser.  I always get the error. See below screenshots.
![result](https://user-images.githubusercontent.com/18308194/45969907-68de5100-c035-11e8-8831-5e19796ef7c6.jpg)
![url screenshot](https://user-images.githubusercontent.com/18308194/45969908-68de5100-c035-11e8-86a8-39eeb7691c59.jpg)

With Microsoft Edge, I see a different error. See below.
![result_edge](https://user-images.githubusercontent.com/18308194/45970045-c83c6100-c035-11e8-81de-adc902f41062.jpg)

Can you share the csv data you used for creating the radar?

**Comment by viveksoundrapandi** [comment]
@sgarap as firefox supports preflight requests for CORS requests, it initiates a preflight request which gets failed as the server[in this case gist.github.com] does not set the response header Access-Control-Allow-Methods explicitly. It only responds with Access-Control-Allow-Origin: * which is not enough for firefox to allow a preflight request.  However, it will work in chrome as pre-flight requests are not released yet .

In short: the server hosting the csv must allow CORS origin and methods in the response headers.

**Comment by mb002270** [comment]
Just worked through this same problem. 

Solution is to go into AWS S3 console (or cli) and alter the CORS settings for the bucket that has your data.  I opened it up with this configuration - which is quite broad...you may want to tighten it down more. 

<CORSConfiguration>
   <CORSRule>
     <AllowedOrigin>*</AllowedOrigin>
     <AllowedMethod>GET</AllowedMethod>
   </CORSRule>
</CORSConfiguration>

See examples at https://docs.aws.amazon.com/AmazonS3/latest/dev/cors.html for more samples

**Comment by viveksoundrapandi** [comment]
@mb002270 thanks this  helps. we will try to add it to the readme to instruct other users as well. closing the issue.

**Comment by sgarap** [comment]
@mb002270 works like a charm. Many Thanks :)

**Comment by gustavz** [comment]
I am not able to get it to work with azure blobs.
There the mandatory CORS options are:
`AllowedOrigins, AllowedMethods, AllowedHeaders, ExposedHeaders, MaxAgeInSeconds`
if I set them to
`*,GET,*,*,99`
It still does not work, any idea why? 

Furthermore @sgarap your example csv `https://s3-eu-west-1.amazonaws.com/tech-radar-eu-west-1/test.csv` only works on `firefox` but does not on `chrome`. Why is that?

Can you help @mb002270?

**Comment by mb002270** [comment]
The browser most likely is doing an OPTIONS request instead of a GET (or maybe even HEAD).  Try adding OPTIONS and GET as valid methods.  If that doesn't work, you can try adding HEAD too.

**Comment by micheldrescher** [comment]
For all I know the radar is doing XmlHttp requests - have you checked that, too?

**Comment by gustavz** [comment]
It finally worked just with GET, the problem was that my csv file had formatting issues.
@mb002270 , @micheldrescher 
What I want to do next is to make it possible to load multiple csv files.
Like using the radar with a google sheet containing multiple sub-sheets they should all be displayed and be able to be selected and visualized separatly.
But as another step further, it should also be possible to select multple of those files and therefore combine them in one visualization.

Do you guys understand what I mean? Furthermore do you know which parts need to be modified to enable this feature?

Thanks again!
